File: leetcode/101.rb
# 101. Symmetric Tree

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}

def is_symmetric(root)
    return true if !root
    return true if (!root.left && !root.right)
    left_child = root.left
    right_child = root.right
    return helper(left_child, right_child)
end

def helper(left, right)
    return true if (!left && !right)
    return false if (!left || !right)
    return false if left.val != right.val
    helper(left.left, right.right) && helper(left.right, right.left)
end

File: leetcode/101.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        queue = deque([root])
        while queue:
            level = []
            for _ in range(len(queue)):
                node = queue.popleft()
                if node:
                    level.append(node.val)
                    queue.append(node.left) if node.left else queue.append(None)
                    queue.append(node.right) if node.right else queue.append(None)
                else:
                    level.append(None)
            i, j = 0, len(level) - 1
            while i <= j:
                if level[i] != level[j]:
                    return False
                i += 1
                j -= 1
        return True